app/web/src/workers/webworker.ts

```diff
@@ -682,32 +705,88 @@ const handleHammer = async (msg: AtomMessage) => {
   });
 };
 
+// Insert atoms in chunks of 2000 per query
+const bulkCreateAtomsFromBulkMjolnir = async (
+  indexObjects: IndexObjectMeta[],
+  chunkSize = 2000,
+) => {
+  for (let i = 0; i < indexObjects.length; i += chunkSize) {
+    const chunk = indexObjects.slice(i, i + chunkSize);
+    const placeholders = [];
+    const bind = [];
+
+    for (const obj of chunk) {
+      placeholders.push("(?, ?, ?, ?)");
+      bind.push(
+        obj.frontEndObject.kind,
+        obj.frontEndObject.checksum,
+        obj.frontEndObject.id,
+        await encodeDocumentForDB(obj.frontEndObject.data),
+      );
+    }
+
+    const sql = `insert into atoms
+        (kind, checksum, args, data)
+          VALUES
+        ${placeholders.join(",")}
+        ON CONFLICT (kind, checksum, args)
+        DO UPDATE SET data=excluded.data;
+      `;
+
+    db.exec({ sql, bind });
+  }
+};
+
+// Insert many-to-many relationships for atoms in chunks of 2000 per query
+const bulkInsertAtomMTMs = (
+  indexObjects: (Common | IndexObjectMeta)[],
+  indexChecksum: Checksum,
+  chunkSize = 2000,
+) => {
+  for (let i = 0; i < indexObjects.length; i += chunkSize) {
+    const chunk = indexObjects.slice(i, i + chunkSize);
+    const placeholders = [];
+    const bind = [];
+    for (const atom of chunk) {
+      placeholders.push("(?, ?, ?, ?)");
+      if ((<IndexObjectMeta>atom).frontEndObject !== undefined) {
+        const obj = atom as IndexObjectMeta;
+        bind.push(
+          indexChecksum,
+          obj.frontEndObject.kind,
+          obj.frontEndObject.id,
+          obj.frontEndObject.checksum,
+        );
+      } else {
+        const obj = atom as Common;
+        bind.push(indexChecksum, obj.kind, obj.id, obj.checksum);
+      }
+    }
+
+    const sql = `insert into index_mtm_atoms
+      (index_checksum, kind, args, checksum)
+      values
+      ${placeholders.join(",")}
+      on conflict (index_checksum, kind, args) do nothing
+      ;`;
+
+    db.exec({ sql, bind });
+  }
+};
+
 const insertAtomMTM = (atom: Atom, indexChecksum: Checksum) => {
   try {
     const bind = [indexChecksum, atom.kind, atom.id, atom.toChecksum];
-    const exists = db.exec({
-      sql: `select index_checksum, kind, args, checksum from index_mtm_atoms
-        where index_checksum = ? and kind = ? and args = ? and checksum = ?
-      ;`,
-      bind,
-      returnValue: "resultRows",
-    });
-    if (exists.length > 0) {
-      return false; // no-op
-    }
-
     db.exec({
       sql: `insert into index_mtm_atoms
         (index_checksum, kind, args, checksum)
           VALUES
         (?, ?, ?, ?)
+        on conflict (index_checksum, kind, args) do nothing
```

Comment:
we _can't_ do this, the table explicitly has a `ON CONFLICT REPLACE`

Reply:
but it works!